File: backend-javascript/src/tenants/tenants.service.ts
import { BadRequestException, Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Tenant } from '../common/entities/tenant.entity';
import { KillSwitchService } from '../common/services/kill-switch.service';
import { TenantAuthService } from '../auth/tenant-auth.service';
import { CreateTenantDto } from './dto/create-tenant.dto';
import { UpdateTenantSelfDto } from './dto/update-tenant-self.dto';
import { UpdateTenantDto } from './dto/update-tenant.dto';

@Injectable()
export class TenantsService {
  constructor(
    @InjectRepository(Tenant)
    private readonly tenantsRepository: Repository<Tenant>,
    private readonly killSwitchService: KillSwitchService,
    private readonly tenantAuthService: TenantAuthService
  ) {}

  private sanitizeTenant(tenant: Tenant) {
    const { authPasswordHash, ...rest } = tenant as Tenant & { authPasswordHash?: string | null };
    return rest;
  }

  async list(tenantId?: string): Promise<Tenant[]> {
    const where = tenantId ? { id: tenantId } : {};
    const tenants = await this.tenantsRepository.find({ where });
    return tenants.map((tenant) => this.sanitizeTenant(tenant) as Tenant);
  }

  async getById(id: string): Promise<Tenant | null> {
    const tenant = await this.tenantsRepository.findOne({ where: { id } });
    return tenant ? (this.sanitizeTenant(tenant) as Tenant) : null;
  }

  async create(dto: CreateTenantDto): Promise<Tenant> {
    if (dto.authUsername) {
      const existing = await this.tenantsRepository.findOne({
        where: { authUsername: dto.authUsername }
      });
      if (existing) {
        throw new BadRequestException('Tenant username already exists');
      }
      if (!dto.authPassword) {
        throw new BadRequestException('Tenant password is required');
      }
    }
    const tenant = this.tenantsRepository.create({
      name: dto.name,
      killSwitch: dto.killSwitch ?? false,
      billingEmail: dto.billingEmail ?? null,
      companyName: dto.companyName ?? null,
      contactName: dto.contactName ?? null,
      phone: dto.phone ?? null,
      addressLine1: dto.addressLine1 ?? null,
      addressLine2: dto.addressLine2 ?? null,
      city: dto.city ?? null,
      postalCode: dto.postalCode ?? null,
      country: dto.country ?? null,
      billingAddressLine1: dto.billingAddressLine1 ?? null,
      billingAddressLine2: dto.billingAddressLine2 ?? null,
      billingCity: dto.billingCity ?? null,
      billingPostalCode: dto.billingPostalCode ?? null,
      billingCountry: dto.billingCountry ?? null,
      taxId: dto.taxId ?? null,
      website: dto.website ?? null,
      authUsername: dto.authUsername ?? null,
      authPasswordHash: dto.authPassword
        ? this.tenantAuthService.buildPasswordHash(dto.authPassword)
        : null,
      authMustChangePassword: dto.authPassword ? true : false
    });
    const saved = await this.tenantsRepository.save(tenant);
    await this.killSwitchService.setTenantKillSwitch(saved.id, saved.killSwitch);
    return this.sanitizeTenant(saved) as Tenant;
  }

  async update(id: string, dto: UpdateTenantDto): Promise<Tenant> {
    const tenant = await this.tenantsRepository.findOne({ where: { id } });
    if (!tenant) {
      throw new NotFoundException('Tenant not found');
    }
    if (dto.authUsername && dto.authUsername !== tenant.authUsername) {
      const existing = await this.tenantsRepository.findOne({
        where: { authUsername: dto.authUsername }
      });
      if (existing) {
        throw new BadRequestException('Tenant username already exists');
      }
    }
    Object.assign(tenant, { ...dto, authPassword: undefined });
    if (dto.authUsername !== undefined) {
      tenant.authUsername = dto.authUsername ?? null;
    }
    if (dto.authPassword) {
      tenant.authPasswordHash = this.tenantAuthService.buildPasswordHash(dto.authPassword);
      tenant.authMustChangePassword = true;
    }
    const saved = await this.tenantsRepository.save(tenant);
    if (dto.killSwitch !== undefined) {
      await this.killSwitchService.setTenantKillSwitch(saved.id, saved.killSwitch);
    }
    return this.sanitizeTenant(saved) as Tenant;
  }

  async updateSelf(id: string, dto: UpdateTenantSelfDto): Promise<Tenant> {
    const tenant = await this.tenantsRepository.findOne({ where: { id } });
    if (!tenant) {
      throw new NotFoundException('Tenant not found');
    }

    Object.assign(tenant, {
      name: dto.name ?? tenant.name,
      billingEmail: dto.billingEmail ?? tenant.billingEmail,
      companyName: dto.companyName ?? tenant.companyName,
      contactName: dto.contactName ?? tenant.contactName,
      phone: dto.phone ?? tenant.phone,
      addressLine1: dto.addressLine1 ?? tenant.addressLine1,
      addressLine2: dto.addressLine2 ?? tenant.addressLine2,
      city: dto.city ?? tenant.city,
      postalCode: dto.postalCode ?? tenant.postalCode,
      country: dto.country ?? tenant.country,
      billingAddressLine1: dto.billingAddressLine1 ?? tenant.billingAddressLine1,
      billingAddressLine2: dto.billingAddressLine2 ?? tenant.billingAddressLine2,
      billingCity: dto.billingCity ?? tenant.billingCity,
      billingPostalCode: dto.billingPostalCode ?? tenant.billingPostalCode,
      billingCountry: dto.billingCountry ?? tenant.billingCountry,
      website: dto.website ?? tenant.website
    });

    const saved = await this.tenantsRepository.save(tenant);
    return this.sanitizeTenant(saved) as Tenant;
  }

  async toggleKillSwitch(id: string, enabled: boolean): Promise<Tenant> {
    return this.update(id, { killSwitch: enabled });
  }
}

File: backend-java/src/main/java/com/neria/manager/tenants/TenantsService.java
package com.neria.manager.tenants;

import com.neria.manager.auth.TenantAuthService;
import com.neria.manager.common.entities.Tenant;
import com.neria.manager.common.repos.TenantRepository;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import org.springframework.stereotype.Service;

@Service
public class TenantsService {
  private final TenantRepository repository;
  private final TenantAuthService tenantAuthService;

  public TenantsService(TenantRepository repository, TenantAuthService tenantAuthService) {
    this.repository = repository;
    this.tenantAuthService = tenantAuthService;
  }

  public List<Tenant> list(String tenantId) {
    if (tenantId != null && !tenantId.isBlank()) {
      return repository.findById(tenantId).map(List::of).orElse(List.of());
    }
    return repository.findAll();
  }

  public Tenant create(CreateTenantRequest dto) {
    Tenant tenant = new Tenant();
    tenant.setId(UUID.randomUUID().toString());
    tenant.setName(dto.name);
    tenant.setStatus(dto.status != null ? dto.status : "active");
    tenant.setKillSwitch(dto.killSwitch != null ? dto.killSwitch : false);
    tenant.setAuthUsername(dto.authUsername);
    tenant.setLanguage("es");
    if (dto.authPassword != null && !dto.authPassword.isBlank()) {
      tenant.setAuthPasswordHash(tenantAuthService.hashPassword(dto.authPassword));
      tenant.setAuthMustChangePassword(true);
    }
    tenant.setCreatedAt(LocalDateTime.now());
    tenant.setUpdatedAt(LocalDateTime.now());
    return repository.save(tenant);
  }

  public Tenant update(String tenantId, UpdateTenantRequest dto) {
    Tenant tenant = repository.findById(tenantId).orElseThrow();
    if (dto.name != null) {
      tenant.setName(dto.name);
    }
    if (dto.status != null) {
      tenant.setStatus(dto.status);
    }
    if (dto.killSwitch != null) {
      tenant.setKillSwitch(dto.killSwitch);
    }
    if (dto.authUsername != null) {
      tenant.setAuthUsername(dto.authUsername);
    }
    if (dto.authPassword != null && !dto.authPassword.isBlank()) {
      tenant.setAuthPasswordHash(tenantAuthService.hashPassword(dto.authPassword));
      tenant.setAuthMustChangePassword(true);
    }
    if (dto.billingEmail != null) {
      tenant.setBillingEmail(dto.billingEmail);
    }
    if (dto.language != null) {
      tenant.setLanguage(dto.language);
    }
    applyProfileFields(tenant, dto);
    tenant.setUpdatedAt(LocalDateTime.now());
    return repository.save(tenant);
  }

  public Tenant updateSelf(String tenantId, UpdateTenantSelfRequest dto) {
    Tenant tenant = repository.findById(tenantId).orElseThrow();
    if (dto.name != null) {
      tenant.setName(dto.name);
    }
    if (dto.billingEmail != null) {
      tenant.setBillingEmail(dto.billingEmail);
    }
    if (dto.language != null) {
      tenant.setLanguage(dto.language);
    }
    if (dto.authPassword != null && !dto.authPassword.isBlank()) {
      tenant.setAuthPasswordHash(tenantAuthService.hashPassword(dto.authPassword));
      tenant.setAuthMustChangePassword(false);
    }
    applyProfileFields(tenant, dto);
    tenant.setUpdatedAt(LocalDateTime.now());
    return repository.save(tenant);
  }

  public Tenant toggleKillSwitch(String tenantId, boolean enabled) {
    Tenant tenant = repository.findById(tenantId).orElseThrow();
    tenant.setKillSwitch(enabled);
    tenant.setUpdatedAt(LocalDateTime.now());
    return repository.save(tenant);
  }

  public Tenant getById(String tenantId) {
    return repository.findById(tenantId).orElse(null);
  }

  private void applyProfileFields(Tenant tenant, UpdateTenantBase dto) {
    if (dto.companyName != null) tenant.setCompanyName(dto.companyName);
    if (dto.contactName != null) tenant.setContactName(dto.contactName);
    if (dto.phone != null) tenant.setPhone(dto.phone);
    if (dto.addressLine1 != null) tenant.setAddressLine1(dto.addressLine1);
    if (dto.addressLine2 != null) tenant.setAddressLine2(dto.addressLine2);
    if (dto.city != null) tenant.setCity(dto.city);
    if (dto.postalCode != null) tenant.setPostalCode(dto.postalCode);
    if (dto.country != null) tenant.setCountry(dto.country);
    if (dto.billingAddressLine1 != null) tenant.setBillingAddressLine1(dto.billingAddressLine1);
    if (dto.billingAddressLine2 != null) tenant.setBillingAddressLine2(dto.billingAddressLine2);
    if (dto.billingCity != null) tenant.setBillingCity(dto.billingCity);
    if (dto.billingPostalCode != null) tenant.setBillingPostalCode(dto.billingPostalCode);
    if (dto.billingCountry != null) tenant.setBillingCountry(dto.billingCountry);
    if (dto.taxId != null) tenant.setTaxId(dto.taxId);
    if (dto.website != null) tenant.setWebsite(dto.website);
    if (dto.language != null) tenant.setLanguage(dto.language);
  }

  public static class UpdateTenantBase {
    public String companyName;
    public String contactName;
    public String phone;
    public String addressLine1;
    public String addressLine2;
    public String city;
    public String postalCode;
    public String country;
    public String billingAddressLine1;
    public String billingAddressLine2;
    public String billingCity;
    public String billingPostalCode;
    public String billingCountry;
    public String taxId;
    public String website;
    public String language;
  }

  public static class CreateTenantRequest extends UpdateTenantBase {
    public String name;
    public String status;
    public Boolean killSwitch;
    public String authUsername;
    public String authPassword;
  }

  public static class UpdateTenantRequest extends UpdateTenantBase {
    public String name;
    public String status;
    public Boolean killSwitch;
    public String authUsername;
    public String authPassword;
    public String billingEmail;
    public String language;
  }

  public static class UpdateTenantSelfRequest extends UpdateTenantBase {
    public String name;
    public String billingEmail;
    public String authPassword;
    public String language;
  }
}
